### ISSUE TITLE
MemberRef

### ISSUE THREAD

**Issue by swankjesse** [square/javapoet#317]
Would be interesting to have an accelerator for member references, like `someObject.instanceMethod()` or `someObject.instanceField` or `SomeEnum.CONSTANT` or `SomeType.staticMethod()` etc.

We'd need a new `MemberRef` type that would include either a TypeName (for static members) or a String name (for instance members).

Would be useful for static imports.

**Comment by jbgi** [comment]
:+1: would be very useful.

**Comment by sormuras** [comment]
What would a possible usage look like? Say, you want to write:
`TimeUnit.DAYS.convert(1701, TimeUnit.SECONDS)`

With `$A` (for anchor like in <a href=...) and `Ref` for the reference type:
```java
"$A(42, $A)", Ref.get(TimeUnit.class, TimeUnit.DAYS, "convert"), Ref.get(TimeUnit.class, TimeUnit.SECONDS)
```
instead of
```java
"$T.$L.convert(42, $T.$L)", TimeUnit.class, TimeUnit.DAYS, TimeUnit.class, TimeUnit.SECONDS
```
or, with `$E` (hint, hint):
```java
"$E.convert(42, $E)", TimeUnit.DAYS, TimeUnit.SECONDS
```

**Comment by jbgi** [comment]
For methods I would want something in the like of ```Ref.get(ExecutableElement method, TypeMirror... typeArguments)```

**Comment by sormuras** [comment]
@jbgi Talking about parameter types, do you? If yes, the parameter types are already included in the `ExecutableElement` instance, iirc.

Anyway, can you give me an example for `someObject.instanceMethod()`. Especially, how do you reference `someObject`. Via a string literal, via NameAllocator?
```java
"$A($A)", Ref.get("this", "equals"), Ref.get(TimeUnit.DAYS)
"$T v = $A", TypeName.get(TimeUnit.class), Ref.get("super", "getCurrentUnit")
```

**Comment by jbgi** [comment]
@sormuras no type arguments. like when I want to generate a call to ```Collections.<String>emptyList()```

**Comment by sormuras** [comment]
I understand. Never used type arguments in that place. Usually the types are infered like:
```java
List<String> strings = Collections.emptyList()
```

**Comment by jbgi** [comment]
@sormuras yes I usually do that. there is cases where type inference is poor however, like when chaining methods:
```Collections.emptyList().stream().findFirst(); // does not compile```
Contrived example, but it does arise sometimes. I agree that the use case is not worth it.

**Comment by sormuras** [comment]
Almost implemented locally... :sweat_smile:

**Comment by sormuras** [comment]
https://gist.github.com/sormuras/f0b8ec30714b2e6f98d9
Your input is much appreciated.

Will clean up and add some javadoc  before creating a PR.

**Comment by sormuras** [comment]
PR #350 is filed.
Some documentation can be previewed here: https://github.com/sormuras/javapoet/tree/ref#r-for-references

**Comment by sormuras** [comment]
See #388 and #404 before implementing `$R` again. :smile:

**Comment by moaxcp** [comment]
Would this be helpful for `switch` statements using an `enum`? I'm running into an issue where when an `enum` is used in a `case` the code does not compile.

```
    byte type = in.readCard8();
    byte data = in.readCard8();
    BehaviorTypeEnum typeEnum = BehaviorTypeEnum.getByCode(type);
    switch(typeEnum) {
      case BehaviorTypeEnum.DEFAULT:
      case BehaviorTypeEnum.LOCK:
      case BehaviorTypeEnum.PERMAMENT_LOCK:
        return new DefaultBehaviorStruct(type);
      case BehaviorTypeEnum.RADIO_GROUP:
      case BehaviorTypeEnum.PERMAMENT_RADIO_GROUP:
        return new RadioGroupBehaviorStruct(type, data);
      case BehaviorTypeEnum.OVERLAY1:
      case BehaviorTypeEnum.OVERLAY2:
      case BehaviorTypeEnum.PERMAMENT_OVERLAY1:
      case BehaviorTypeEnum.PERMAMENT_OVERLAY2:
        return new OverlayBehaviorStruct(type, data);
      default:
        return new CommonBehaviorStruct(type, data);
    }
```

This is the error

```
/home/john/projects/x11-client/build/generated/xproto/com/github/moaxcp/x11client/protocol/xkb/BehaviorUnion.java:13: error: an enum switch case label must be the unqualified name of an enumeration constant
      case BehaviorTypeEnum.DEFAULT:
                           ^
```

In this case the `enum` value must be statically imported.

**Comment by ronshapiro** [comment]
I don't believe you need to statically import here. javac knows to look for
the enum values directly without the import.

בתאריך יום א׳, 16 באוג׳ 2020, 1:05, מאת John Mercier ‏<
notifications@github.com>:

**Comment by moaxcp** [comment]
I don't think it does. When I manually edit the code to static import the error goes away. This seems like a common problem developers run into when using enums with a switch. I have found many other examples online for this error message.

I think this makes a good argument for implementing $R and requesting static imports from a CodeBlock.
